src/operator/keycloak.ts

```diff
@@ -601,178 +552,149 @@ async function internalIdp(api: KeycloakApi, connection: KeycloakConnection) {
-    groups.realmGroupsGet(keycloakRealm),
-  )) || []) as Array<GroupRepresentation>
+  console.info('Getting realm groups')
   try {
+    const existingGroups = ((await api.groups.realmGroupsGet(keycloakRealm)).body || []) as GroupRepresentation[]
     await Promise.all(
       teamGroups.map((group) => {
         const groupName = group.name!
         const existingGroup = existingGroups.find((el) => el.name === groupName)
         if (existingGroup) {
-          return doApiCall(errors, `Updating groups ${groupName}`, async () =>
-            groups.realmGroupsIdPut(keycloakRealm, existingGroup.id!, group),
-          )
+          console.info(`Updating groups ${groupName}`)
+          return api.groups.realmGroupsIdPut(keycloakRealm, existingGroup.id!, group)
         }
-        return doApiCall(errors, `Creating group ${groupName}`, async () =>
-          groups.realmGroupsPost(keycloakRealm, group),
-        )
+        console.info(`Creating group ${groupName}`)
+        return api.groups.realmGroupsPost(keycloakRealm, group)
       }),
     )
     console.info('Finished managing groups')
-    await keycloakConfigMapChanges()
   } catch (error) {
     throw extractError('managing groups', error)
   }
 }
 
-export async function removeUserGroups(
-  api: { users: UsersApi; groups: GroupsApi },
-  existingUser: UserRepresentation,
-  teamGroups: string[],
-): Promise<void> {
-  try {
-    const { body: existingUserGroups } = await api.users.realmUsersIdGroupsGet(keycloakRealm, existingUser.id as string)
-
-    await Promise.all(
-      existingUserGroups.map(async (group) => {
-        if (!teamGroups.includes(group.name)) {
-          await api.users.realmUsersIdGroupsGroupIdDelete(keycloakRealm, existingUser.id as string, group.id as string)
-        }
-      }),
-    )
-  } catch (error) {
-    throw extractError('removing user groups', error)
-  }
-}
-
-export async function addUserGroups(
-  api: { users: UsersApi; groups: GroupsApi },
-  existingUser: UserRepresentation,
+export async function updateUserGroups(
+  api: KeycloakApi,
+  user: UserRepresentation,
+  groupsByName: Record<string, string>,
   teamGroups: string[],
 ): Promise<void> {
-  try {
-    const { body: currentKeycloakGroups } = await api.groups.realmGroupsGet(keycloakRealm)
-    const { body: existingUserGroups } = await api.users.realmUsersIdGroupsGet(keycloakRealm, existingUser.id as string)
-
-    await Promise.all(
-      teamGroups.map(async (teamGroup) => {
-        const existingGroup = existingUserGroups.find((el) => el.name === teamGroup)
-
-        if (!existingGroup) {
-          const groupId = currentKeycloakGroups.find((el) => el.name === teamGroup)?.id
-          if (groupId) {
-            await api.users.realmUsersIdGroupsGroupIdPut(keycloakRealm, existingUser.id as string, groupId as string)
-          }
+  const userId = user.id!
+  const existingUserGroups = (await api.users.realmUsersIdGroupsGet(keycloakRealm, userId)).body
```

Comment:
I think destructing is the more preferable approach than chaining in typescript. As it is clearer distinction between the api call and what we extract the body
```suggestion
  const { body: existingUserGroups } = (await api.users.realmUsersIdGroupsGet(keycloakRealm, userId))
```

Reply:
I would agree, but unfortunately it does not seem possible to introduce a type during deconstruction. And the API library we work with is not typed, it just returns objects. Therefore we either have to do the type casting separately, or would have to write another wrapper around it. Or just work with Object from here, but I would clearly prefer not to do that.

Reply:
We don't introduce a type here right? So it should be possible, I think